### PR TITLE
sort and deduplicate story events before adding them to NLU training data

### PR DESCRIPTION
**Pull request by wochinge**
**Proposed changes**:
- fix https://github.com/RasaHQ/rasa/issues/6659
- A change of the story order or the order within the steps should not affect the NLU training data. If this constraint is not respected, then the NLU model will be retrained whenever the story or event order changed. To guarantee the stableness of the NLU data, the events are deduplicated and ordered before generating the `Message` objects from them. Note: New actions / utterances still need to cause a new training of the NLU model as this is required for the end-to-end support
- get rid of a couple of `asyncio.coroutine` usages as this function is deprecated as of Python 3.8

**Status (please check what you already did)**:
- [x] added some tests for the functionality
- [ ] updated the documentation
- [ ] updated the changelog (please check [changelog](https://github.com/RasaHQ/rasa/tree/master/changelog) for instructions)
- [x] reformat files using `black` (please check [Readme](https://github.com/RasaHQ/rasa#code-style) for instructions)

### CODE REVIEW

File: rasa/shared/importers/importer.py [hunk 1]
```diff
@@ -488,12 +488,19 @@ async def get_nlu_data(self, language: Optional[Text] = "en") -> TrainingData:
...
-                    additional_messages_from_stories.append(message)
+        utterances, actions = _unique_events_from_stories(stories)
+
+        # Sort events to guarantee deterministic behavior and to avoid that the NLU
+        # model has to be retrained due to changes in the event order within
+        # the stories.
+        sorted_utterances = sorted(
+            list(utterances), key=lambda user: user.intent_name or user.text
+        )
+        sorted_actions = sorted(list(actions), key=lambda action: action.action_name)
```
Comment (evgeniiaraz) [inline rasa/shared/importers/importer.py:499]: why are we not using `action_text` here?

File: rasa/shared/importers/importer.py [hunk 2]
```diff
@@ -488,12 +488,19 @@ async def get_nlu_data(self, language: Optional[Text] = "en") -> TrainingData:
...
-                message = _message_from_conversation_event(event)
-                if message:
-                    additional_messages_from_stories.append(message)
+        utterances, actions = _unique_events_from_stories(stories)
+
+        # Sort events to guarantee deterministic behavior and to avoid that the NLU
+        # model has to be retrained due to changes in the event order within
+        # the stories.
+        sorted_utterances = sorted(
+            list(utterances), key=lambda user: user.intent_name or user.text
```
Comment (ricwo) [inline rasa/shared/importers/importer.py:497]:
```suggestion
            utterances, key=lambda user: user.intent_name or user.text
```

File: rasa/shared/importers/importer.py (hunk 1 above)
Comment (ricwo) [inline rasa/shared/importers/importer.py:499]:
```suggestion
        sorted_actions = sorted(actions, key=lambda action: action.action_name)
```

File: tests/shared/importers/test_importer.py [hunk 3]
```diff
@@ -228,40 +238,69 @@ async def test_import_nlu_training_data_from_e2e_stories(project: Text):
...
+    # Patch to return our test stories
+    default_importer.importer.get_stories = mocked_stories
+
+    training_data = await default_importer.get_nlu_data()
+
+    # Pretend the order of  the stories changed. This should have no
+    # effect on the NLU training data
+    stories = list(reversed(stories))
+
+    # Make sure importer doesnt cache stories
```
Comment (ricwo) [inline tests/shared/importers/test_importer.py:274]:
```suggestion
    # Make sure importer doesn't cache stories
```

**Comment by evgeniiaraz** [comment]
Just asking for my understanding -- why `rasa.nlu.training_data.training_data.sanitize_examples` isn't enough? I though it did the ordering by turning training examples into ordered dict. Or are we by-passing the initialization somehow?

**Review by ricwo — APPROVED** [review]
Looks great 👍

File: rasa/shared/importers/importer.py (hunk 1 above)
Comment (wochinge) [reply rasa/shared/importers/importer.py]: I was under the impression that always `action_name==action_text` . Added it now 👍

Comment (wochinge) [reply rasa/shared/importers/importer.py]: Makes sense that `sorted` returns a list 🙈

File: rasa/shared/importers/importer.py [hunk 4]
```diff
@@ -496,7 +496,9 @@ async def _additional_training_data_from_stories(self) -> TrainingData:
         sorted_utterances = sorted(
             utterances, key=lambda user: user.intent_name or user.text
         )
-        sorted_actions = sorted(actions, key=lambda action: action.action_name)
+        sorted_actions = sorted(
+            actions, key=lambda action: action.action_name or action.action_text
```
Comment (ricwo) [inline rasa/shared/importers/importer.py:500]: but an action name can't be `None` or `''` can it (unlike the intent name)?

Comment (wochinge) [reply rasa/shared/importers/importer.py:500]: it will actually once we add end-to-end reading 👍 (see https://github.com/RasaHQ/rasa/pull/6605)

**Pull request merged**